### PR TITLE
Use fact() function for all os.distro.* facts

### PR DESCRIPTION
**Pull request by root-expert**
* On Puppet 6 facter 3.x requires `lsb-release` to resolve os.distro.* facts which may not always be available. Using $facts hash cause errors like "Evaluation Error: Operator '[]' is not applicable to an Undef Value." because os.distro is undefined causing the catalog to fail. Use fact() to identify Undef facts and throw an error to the user.

Releated:

* https://github.com/voxpupuli/beaker-hostgenerator/pull/244

### CODE REVIEW

**Review by kenyon — DISMISSED** [review]

**Review by smortex — COMMENTED** [review]
Hum… :thinking:  This is a good start but not enough I think.

This is replacing a hard failure with a cryptic error message by no failure (at the puppet level) but a broken configuration (at the apt level) when running on a system without the 'apt-transport-https' package but 'https://' repositories (and all sources will lack a release which is probably not good BTW).

We can't add a dependency on `lsb-release` at this level since it is needed before the catalog is computed (i.e. already too late).

My guess is that we should rather fail hard with a clear error message when we detect that `fact('os.distro.codename')` is `undef` with a message like `Cannot determine the distribution codename. Older versions of Facter rely on lsb-release to fetch it, is it installed?`.  Maybe in the `apt` class if all classes include it.

What do you think?

**Comment by root-expert** [comment]
Yes that's probably the correct way and I have opened a PR on the puppetlabs-apt repo but nobody has gotten around it just yet.

I've seen some more modules using os.distro facts but the good thing is they are limited.

**Comment by root-expert** [comment]
I don't think it's a cryptic error, it clearly states that a specific parameter must be declared in order for it to work 🤔 

It fails to auto-detect the distro so the user must explicitly set it (or install lsb-release)

**Comment by smortex** [comment]
What I call a *cryptic* error message is [the following when adding a random resource](https://github.com/voxpupuli/puppet-elasticsearch/runs/5374994344?check_suite_focus=true#step:4:599):
```
  Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Operator '[]' is not applicable to an Undef Value. (file: /etc/puppetlabs/code/environments/production/modules/apt/manifests/source.pp, line: 102, column: 9) (file: /etc/puppetlabs/code/environments/production/modules/elastic_stack/manifests/repo.pp, line: 71) on node ubuntu1804-64.example.com
```
While the root problem is that a package is missing.

As it is right now, the PR fix the catalog but I can imagine cases where this will produce broken apt config.  I did a test for the above module but it seems to be okay (https://github.com/voxpupuli/puppet-elasticsearch/pull/1149) so maybe I am paranoid? :shrug: :smile:

**Comment by kenyon** [comment]
@smortex in that elasticsearch test run, why is it complaining about source.pp line 102? That's just a variable assignment. 🤔

**Comment by smortex** [comment]
> @smortex in that elasticsearch test run, why is it complaining about source.pp line 102? That's just a variable assignment. thinking

Line 102 in [the latest release](https://github.com/puppetlabs/puppetlabs-apt/blob/v8.3.0/manifests/source.pp#L102), but line 103 on `main` ATM:

```puppet
if ($facts['os']['distro']['codename'] in $_transport_https_releases) and $_location =~ /(?i:^https:\/\/)/ {
```

Without `lsb-release`, with facter 3, `$facts['os']['distro']` is `undef`

**Comment by root-expert** [comment]
> > @smortex in that elasticsearch test run, why is it complaining about source.pp line 102? That's just a variable assignment. thinking
> 
> Line 102 in [the latest release](https://github.com/puppetlabs/puppetlabs-apt/blob/v8.3.0/manifests/source.pp#L102), but line 103 on `main` ATM:
> 
> ```puppet
> if ($facts['os']['distro']['codename'] in $_transport_https_releases) and $_location =~ /(?i:^https:\/\/)/ {
> ```
> 
> Without `lsb-release`, with facter 3, `$facts['os']['distro']` is `undef`

Ah I see you point. Yea that was the only part which the module doesn't fail. Failing directly on the `apt` class maybe it's too harsh? I would maybe opt out to fail per-class although not a strong opinion here.

File: manifests/source.pp [hunk 1]
```diff
@@ -99,8 +99,8 @@
       $_location = $location
     }
     # Newer oses, do not need the package for HTTPS transport.
-    $_transport_https_releases = [ 'wheezy', 'jessie', 'stretch', 'trusty', 'xenial' ]
-    if ($facts['os']['distro']['codename'] in $_transport_https_releases) and $_location =~ /(?i:^https:\/\/)/ {
+    $_transport_https_releases = [ '7', '8', '9', '14.04', '16.04' ]
+    if (fact('os.relese.major') in $_transport_https_releases) and $_location =~ /(?i:^https:\/\/)/ {
```
Comment (ekohl) [inline manifests/source.pp:103]:
```suggestion
    if (fact('os.release.major') in $_transport_https_releases) and $_location =~ /(?i:^https:\/\/)/ {
```

File: manifests/source.pp [hunk 2]
```diff
@@ -79,8 +79,8 @@
   $_before = Apt::Setting["list-${title}"]
 
   if !$release {
-    if $facts['os']['distro']['codename'] {
-      $_release = $facts['os']['distro']['codename']
+    if fact('os.distro.codename') {
+      $_release = fact('os.distro.codename')
```
Comment (ekohl) [inline manifests/source.pp:83]: Very strictly speaking, this line doesn't need to be modified since at this point you know it's available. But it works too.

**Review by ekohl — CHANGES_REQUESTED** [review]
Found a typo, other than that :+1:

File: manifests/source.pp (hunk 1 above)
Comment (root-expert) [reply manifests/source.pp:103]: Thanks!

**Review by ekohl — APPROVED** [review]

**Comment by ekohl** [comment]
A few times I've forgotten to merge after approving but having to wait for the tests. If I do, please ping me.

**Review by smortex — APPROVED** [review]
I am fine with this: maybe some edge case still need adjusting but it proves to improve the situation :+1:
